File: auth0/src/main/java/com/auth0/android/request/internal/BaseAuthenticationRequest.java
package com.auth0.android.request.internal;

import android.util.Log;

import androidx.annotation.NonNull;

import com.auth0.android.authentication.AuthenticationException;
import com.auth0.android.request.AuthenticationRequest;
import com.auth0.android.result.Credentials;
import com.google.gson.Gson;
import com.squareup.okhttp.HttpUrl;
import com.squareup.okhttp.OkHttpClient;

import java.util.HashMap;
import java.util.Map;

import static com.auth0.android.authentication.ParameterBuilder.AUDIENCE_KEY;
import static com.auth0.android.authentication.ParameterBuilder.CONNECTION_KEY;
import static com.auth0.android.authentication.ParameterBuilder.DEVICE_KEY;
import static com.auth0.android.authentication.ParameterBuilder.GRANT_TYPE_KEY;
import static com.auth0.android.authentication.ParameterBuilder.REALM_KEY;
import static com.auth0.android.authentication.ParameterBuilder.SCOPE_KEY;

class BaseAuthenticationRequest extends SimpleRequest<Credentials, AuthenticationException> implements AuthenticationRequest {

    private static final String TAG = BaseAuthenticationRequest.class.getSimpleName();

    public BaseAuthenticationRequest(HttpUrl url, OkHttpClient client, Gson gson, String httpMethod) {
        super(url, client, gson, httpMethod, Credentials.class, new AuthenticationErrorBuilder());
    }

    private boolean hasLegacyPath() {
        return !url.encodedPath().equals("/oauth/token");
    }

    /**
     * Sets the 'grant_type' parameter
     *
     * @param grantType grant type
     * @return itself
     */
    @NonNull
    @Override
    public AuthenticationRequest setGrantType(@NonNull String grantType) {
        addParameter(GRANT_TYPE_KEY, grantType);
        return this;
    }

    /**
     * Sets the 'connection' parameter.
     *
     * @param connection name of the connection
     * @return itself
     */
    @NonNull
    @Override
    public AuthenticationRequest setConnection(@NonNull String connection) {
        if (!hasLegacyPath()) {
            Log.w(TAG, "Not setting the 'connection' parameter as the request is using a OAuth 2.0 API Authorization endpoint that doesn't support it.");
            return this;
        }
        addParameter(CONNECTION_KEY, connection);
        return this;
    }

    /**
     * Sets the 'realm' parameter. A realm identifies the host against which the authentication will be made, and usually helps to know which username and password to use.
     *
     * @param realm name of the realm
     * @return itself
     */
    @NonNull
    @Override
    public AuthenticationRequest setRealm(@NonNull String realm) {
        if (hasLegacyPath()) {
            Log.w(TAG, "Not setting the 'realm' parameter as the request is using a Legacy Authorization API endpoint that doesn't support it.");
            return this;
        }
        addParameter(REALM_KEY, realm);
        return this;
    }

    /**
     * Sets the 'scope' parameter.
     *
     * @param scope a scope value
     * @return itself
     */
    @NonNull
    public AuthenticationRequest setScope(@NonNull String scope) {
        addParameter(SCOPE_KEY, scope);
        return this;
    }

    /**
     * Sets the 'device' parameter
     *
     * @param device a device name
     * @return itself
     */
    @NonNull
    public AuthenticationRequest setDevice(@NonNull String device) {
        addParameter(DEVICE_KEY, device);
        return this;
    }

    /**
     * Sets the 'audience' parameter.
     *
     * @param audience an audience value
     * @return itself
     */
    @NonNull
    @Override
    public AuthenticationRequest setAudience(@NonNull String audience) {
        addParameter(AUDIENCE_KEY, audience);
        return this;
    }

    @NonNull
    @Override
    public AuthenticationRequest addAuthenticationParameters(@NonNull Map<String, Object> parameters) {
        final HashMap<String, Object> params = new HashMap<>(parameters);
        if (parameters.containsKey(CONNECTION_KEY)) {
            setConnection((String) params.remove(CONNECTION_KEY));
        }
        if (parameters.containsKey(REALM_KEY)) {
            setRealm((String) params.remove(REALM_KEY));
        }
        addParameters(params);
        return this;
    }

    @NonNull
    @Override
    public BaseAuthenticationRequest addHeader(@NonNull String name, @NonNull String value) {
        super.addHeader(name, value);
        return this;
    }
}

File: auth0/src/main/java/com/auth0/android/request/internal/BaseAuthenticationRequest.kt
package com.auth0.android.request.internal

import android.text.TextUtils
import android.util.Log
import androidx.annotation.VisibleForTesting
import com.auth0.android.Auth0Exception
import com.auth0.android.authentication.AuthenticationException
import com.auth0.android.authentication.ParameterBuilder
import com.auth0.android.callback.Callback
import com.auth0.android.provider.IdTokenMissingException
import com.auth0.android.provider.IdTokenVerificationOptions
import com.auth0.android.provider.IdTokenVerifier
import com.auth0.android.provider.TokenValidationException
import com.auth0.android.provider.UnexpectedIdTokenException
import com.auth0.android.request.AuthenticationRequest
import com.auth0.android.request.Request
import com.auth0.android.request.RequestValidator
import com.auth0.android.result.Credentials
import java.util.Date

internal open class BaseAuthenticationRequest(
    private val request: Request<Credentials, AuthenticationException>,
    private val clientId: String, baseURL: String
) : AuthenticationRequest {

    private companion object {
        private val TAG = BaseAuthenticationRequest::class.java.simpleName
        private const val ERROR_VALUE_ID_TOKEN_VALIDATION_FAILED = "Could not verify the ID token"
    }

    private var _currentTimeInMillis: Long? = null

    @VisibleForTesting(otherwise = VisibleForTesting.PRIVATE)
    internal var validateClaims = false

    @VisibleForTesting(otherwise = VisibleForTesting.PRIVATE)
    internal var idTokenVerificationLeeway: Int? = null

    @VisibleForTesting(otherwise = VisibleForTesting.PRIVATE)
    internal var idTokenVerificationIssuer: String = baseURL

    @set:VisibleForTesting(otherwise = VisibleForTesting.PRIVATE)
    internal var currentTimeInMillis: Long
        get() = if (_currentTimeInMillis != null) _currentTimeInMillis!! else System.currentTimeMillis()
        set(value) {
            _currentTimeInMillis = value
        }


    /**
     * Sets the 'grant_type' parameter
     *
     * @param grantType grant type
     * @return itself
     */
    override fun setGrantType(grantType: String): AuthenticationRequest {
        addParameter(ParameterBuilder.GRANT_TYPE_KEY, grantType)
        return this
    }

    /**
     * Sets the 'connection' parameter.
     *
     * @param connection name of the connection
     * @return itself
     */
    override fun setConnection(connection: String): AuthenticationRequest {
        addParameter(ParameterBuilder.CONNECTION_KEY, connection)
        return this
    }

    /**
     * Sets the 'realm' parameter. A realm identifies the host against which the authentication will be made, and usually helps to know which username and password to use.
     *
     * @param realm name of the realm
     * @return itself
     */
    override fun setRealm(realm: String): AuthenticationRequest {
        addParameter(ParameterBuilder.REALM_KEY, realm)
        return this
    }

    /**
     * Sets the 'scope' parameter.
     *
     * @param scope a scope value
     * @return itself
     */
    override fun setScope(scope: String): AuthenticationRequest {
        addParameter(ParameterBuilder.SCOPE_KEY, scope)
        return this
    }

    /**
     * Sets the 'audience' parameter.
     *
     * @param audience an audience value
     * @return itself
     */
    override fun setAudience(audience: String): AuthenticationRequest {
        addParameter(ParameterBuilder.AUDIENCE_KEY, audience)
        return this
    }

    override fun addValidator(validator: RequestValidator): AuthenticationRequest {
        request.addValidator(validator)
        return this
    }

    override fun validateClaims(): AuthenticationRequest {
        this.validateClaims = true
        return this
    }

    override fun withIdTokenVerificationLeeway(leeway: Int): AuthenticationRequest {
        this.idTokenVerificationLeeway = leeway
        return this
    }

    override fun withIdTokenVerificationIssuer(issuer: String): AuthenticationRequest {
        this.idTokenVerificationIssuer = issuer
        return this
    }

    override fun addParameters(parameters: Map<String, String>): AuthenticationRequest {
        request.addParameters(parameters)
        return this
    }

    override fun addParameter(name: String, value: String): AuthenticationRequest {
        request.addParameter(name, value)
        return this
    }

    override fun addParameter(name: String, value: Any): AuthenticationRequest {
        request.addParameter(name, value)
        return this
    }

    override fun addHeader(name: String, value: String): AuthenticationRequest {
        request.addHeader(name, value)
        return this
    }

    override fun start(callback: Callback<Credentials, AuthenticationException>) {
        warnClaimValidation()
        request.start(object : Callback<Credentials, AuthenticationException> {
            override fun onSuccess(result: Credentials) {
                if (validateClaims) {
                    try {
                        verifyClaims(result.idToken)
                    } catch (e: AuthenticationException) {
                        callback.onFailure(e)
                        return
                    }
                }
                callback.onSuccess(result)
            }

            override fun onFailure(error: AuthenticationException) {
                callback.onFailure(error)
            }
        })
    }

    @Throws(Auth0Exception::class)
    override fun execute(): Credentials {
        warnClaimValidation()
        val credentials = request.execute()
        if (validateClaims) {
            verifyClaims(credentials.idToken)
        }
        return credentials
    }

    @JvmSynthetic
    @Throws(Auth0Exception::class)
    override suspend fun await(): Credentials {
        warnClaimValidation()
        val credentials = request.await()
        if (validateClaims) {
            verifyClaims(credentials.idToken)
        }
        return credentials
    }

    /**
     * Used to verify the claims from the ID Token.
     *
     * @param idToken - The ID Token obtained through authentication
     * @throws AuthenticationException - This is a exception wrapping around [TokenValidationException]
     */
    @VisibleForTesting(otherwise = VisibleForTesting.PRIVATE)
    internal fun verifyClaims(idToken: String) {
        try {
            if (TextUtils.isEmpty(idToken)) {
                throw IdTokenMissingException()
            }
            val decodedIdToken: Jwt = try {
                Jwt(idToken)
            } catch (error: Exception) {
                throw UnexpectedIdTokenException(error)
            }
            val options = IdTokenVerificationOptions(
                idTokenVerificationIssuer,
                clientId,
                null
            )
            options.clockSkew = idTokenVerificationLeeway
            options.clock = Date(currentTimeInMillis)
            IdTokenVerifier().verify(decodedIdToken, options, false)
        } catch (e: TokenValidationException) {
            throw AuthenticationException(ERROR_VALUE_ID_TOKEN_VALIDATION_FAILED, e)
        }
    }

    private fun warnClaimValidation() {
        if (!validateClaims) {
            Log.e(
                TAG,
                "The request is made without validating claims. Enable claim validation by calling AuthenticationRequest#validateClaims()"
            )
        }
    }
}